src/nvim/os/job.c

```diff
@@ -55,6 +56,22 @@ uv_signal_t schld;
 #endif
 // Callbacks for libuv
 
+static int count = 0;  // A cyclical counter used for assigning job IDs.
+
+/// Returns a new job ID within the range `[1, INT_MAX]`.
+///
+/// This number will be unique between all jobs that have not been stopped, but
+/// cycle back to one after reaching `INT_MAX`. This makes it highly unlikely
+/// for two jobs to have the same ID.
+static int new_job_id(void)
+{
+  count = count < INT_MAX ? count + 1 : 1;
```

Comment:
@oni-link: From `:h jobstart`:

>		Returns:
>		  - The job ID on success, which is used by |jobsend()| and
>		    |jobstop()|
>		  - 0 when the job table is full or on invalid arguments
>		  - -1 when {cmd}[0] is not executable.  Will never fail if
>		    {cmd} is a string unless 'shell' is not executable.

. 0 and -1 are reserved. If you simply use `unsigned` then you will be able to have *both* values because in VimL there are only signed integers.